wapitiCore/attack/mod_csp.py

```diff
@@ -21,16 +21,68 @@
+from wapitiCore.net.csp_utils import csp_header_to_dict, CSP_CHECK_LISTS
 from wapitiCore.definitions.csp import NAME, WSTG_CODE
 from wapitiCore.main.log import log_red
 
 MSG_NO_CSP = "CSP is not set"
-MSG_CSP_MISSING = "CSP attribute \"{0}\" is missing"
-MSG_CSP_UNSAFE = "CSP \"{0}\" value is not safe"
 
+INFO_UNSAFE_INLINE = "\"unsafe-inline\" in \"{0}\" directive allows the execution of unsafe in-page scripts and event\
+ handlers."
+INFO_UNSAFE_EVAL = "\"unsafe-eval\" in \"{0}\" directive allows the execution of code injected into DOM APIs such as\
+ eval()."
+INFO_DATA_HTTP_HTTPS = "value \"{0}\" URI in \"{1}\" allows the execution of unsafe scripts."
+INFO_ALLOW_ALL = "\"{0}\" directive should not allow \"*\" as source"
+INFO_UNSAFE_OBJECT_SRC = "unsafe values \"{0}\" other then \"none\" identified in \"object-src\""
+INFO_UNSAFE_BASE_URI = "unsafe values \"{0}\" other then \"none\" and \"self\" identified in \"base-uri\""
+INFO_UNDEFINED_DIRECTIVE = "directive \"{0}\" is not defined"
 
 # This module check the basics recommendations of CSP
+def check_policy(policy_name, csp_dict):
+    """
+    This function return the unsafe values for each directive of the tested CSP
+    """
+    info = ""
+
+    if policy_name not in csp_dict and "default-src" not in csp_dict:
+        log_red(INFO_UNDEFINED_DIRECTIVE.format(policy_name))
+        info += INFO_UNDEFINED_DIRECTIVE.format(policy_name) + "\n"
+
+    # The HTTP CSP "default-src" directive serves as a fallback for the other CSP fetch directives.
+    # https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/default-src
+    policy_values = csp_dict.get(policy_name) or csp_dict["default-src"]
+    info = ""
+    # If the tested element is default-src or script-src, we must ensure that none of this unsafe values are present
+    if policy_name in ["default-src", "script-src"]:
+        for unsafe_value in CSP_CHECK_LISTS[policy_name]:
+            if unsafe_value in policy_values:
+                if unsafe_value == "unsafe-inline":
+                    log_red(INFO_UNSAFE_INLINE.format(policy_name))
+                    info += INFO_UNSAFE_INLINE.format(policy_name) + "\n"
+                elif unsafe_value in ("data:", "http:", "https:"):
+                    log_red(INFO_DATA_HTTP_HTTPS.format(unsafe_value, policy_name))
+                    info += INFO_DATA_HTTP_HTTPS.format(unsafe_value, policy_name) + "\n"
+                elif unsafe_value == "*":
+                    log_red(INFO_ALLOW_ALL.format(policy_name))
+                    info += INFO_ALLOW_ALL.format(policy_name) + "\n"
+                elif unsafe_value == "unsafe-eval":
+                    log_red(INFO_UNSAFE_EVAL.format(policy_name))
+                    info += INFO_UNSAFE_EVAL.format(policy_name) + "\n"
+
+    # If the tested element is none of the previous list, we must ensure that one of this safe values is present
+    else:
+        for safe_value in CSP_CHECK_LISTS[policy_name]:
```

Comment:
That loop is wrong as it doesn't do the action described in comment.

For example when tested on github.com it displays 

`unsafe values "['self']" other then "none" and "self" identified in "base-uri"`

while github has the policy `base-uri 'self'`

Instead the loop should iterate over `policy_values` THEN check if the value is inside `CSP_CHECK_LISTS[policy_name]`

However on case like  `base-uri yolo 'self'` it will still display something while `self` is in the policy.

I think the best way would be to use sets like this:

```python
if not set(policy_values) & set(CSP_CHECK_LISTS[policy_name]):
  // log policy values
```

Also `policy_values` should be translated to a string joined by commas when displayed.